File: core/metadata/src/main/java/io/theficos/ereader/core/metadata/Isbn.kt
package io.theficos.ereader.core.metadata

object Isbn {
    fun toIsbn13(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        val s = raw.trim().replace("-", "").replace(" ", "").uppercase()
        if (isIsbn13(s)) return s
        if (isIsbn10(s)) return isbn10to13(s)
        return null
    }

    private fun isIsbn13(s: String): Boolean {
        if (s.length != 13 || !s.all { it.isDigit() }) return false
        val sum = s.mapIndexed { i, c -> (if (i % 2 == 0) 1 else 3) * (c - '0') }.sum()
        return sum % 10 == 0
    }

    private fun isIsbn10(s: String): Boolean {
        if (s.length != 10) return false
        var total = 0
        for (i in 0 until 10) {
            val c = s[i]
            val v = when {
                c == 'X' && i == 9 -> 10
                c.isDigit() -> c - '0'
                else -> return false
            }
            total += (10 - i) * v
        }
        return total % 11 == 0
    }

    private fun isbn10to13(s: String): String {
        val core = "978" + s.substring(0, 9)
        val check = (10 - core.mapIndexed { i, c -> (if (i % 2 == 0) 1 else 3) * (c - '0') }.sum() % 10) % 10
        return core + check
    }
}

File: server/quire_server/core/isbn.py
"""ISBN normalization: validate ISBN-10/13 and canonicalize to ISBN-13.

A single source of truth so scanned/typed ISBNs are compared and keyed
consistently (owned-detection, identity.metadata_id="isbn:<isbn13>").
"""

from __future__ import annotations


def _clean(raw: str) -> str:
    return raw.strip().replace("-", "").replace(" ", "").upper()


def _isbn10_valid(s: str) -> bool:
    if len(s) != 10:
        return False
    total = 0
    for i, ch in enumerate(s):
        if ch == "X" and i == 9:
            val = 10
        elif ch.isdigit():
            val = int(ch)
        else:
            return False
        total += (10 - i) * val
    return total % 11 == 0


def _isbn13_valid(s: str) -> bool:
    if len(s) != 13 or not s.isdigit():
        return False
    total = sum((1 if i % 2 == 0 else 3) * int(c) for i, c in enumerate(s))
    return total % 10 == 0


def _isbn10_to_13(s: str) -> str:
    core = "978" + s[:9]
    check = (10 - sum((1 if i % 2 == 0 else 3) * int(c) for i, c in enumerate(core)) % 10) % 10
    return core + str(check)


def to_isbn13(raw: str | None) -> str | None:
    """Return the canonical ISBN-13 for a valid ISBN-10/13, else None."""
    if not raw:
        return None
    s = _clean(raw)
    if _isbn13_valid(s):
        return s
    if _isbn10_valid(s):
        return _isbn10_to_13(s)
    return None
